src/net_processing.cpp

```diff
@@ -1293,31 +1313,44 @@ void PeerManagerImpl::FindNextBlocksToDownload(const Peer& peer, unsigned int co
     // Make sure pindexBestKnownBlock is up to date, we'll need it.
     ProcessBlockAvailability(peer.m_id);
 
-    if (state->pindexBestKnownBlock == nullptr || state->pindexBestKnownBlock->nChainWork < m_chainman.ActiveChain().Tip()->nChainWork || state->pindexBestKnownBlock->nChainWork < m_chainman.MinimumChainWork()) {
+    const CChain& our_chain = chainstate->m_chain;
+    const CBlockIndex* our_tip = our_chain.Tip();
+
+    if (state->pindexBestKnownBlock == nullptr
+            || state->pindexBestKnownBlock->nChainWork < our_tip->nChainWork
+            || state->pindexBestKnownBlock->nChainWork < m_chainman.MinimumChainWork()) {
         // This peer has nothing interesting.
         return;
     }
 
-    if (state->pindexLastCommonBlock == nullptr) {
+    if (!state->chainstate_to_last_common_block.count(chainstate)) {
```

Comment:
There's a lot of map lookups in this function, and it seems like you could eliminate them all by declaring a local reference

```
const CBlockIndex*& last_common_block =  state->chainstate_to_last_common_block[chainstate];
```

Note the pointer is const but the reference is non-const, so you could assign to it and make fewer changes to code below, just replacing `state->pindexLastCommonBlock` with `last_common_block` and `m_chainman.ActiveChain()` with `our_chain`.